### ISSUE TITLE
Kafka initialization failed due to injection of OpenTelemetryMetricsReporter.

### ISSUE THREAD

**Issue by Cirilla-zmh** [open-telemetry/opentelemetry-java-instrumentation#12538]
When we install OTel Java Agent for an application that integrates the Kafka client, it causes the Kafka producer initialization to fail. We eventually identified and resolved the issue, but I believe it's worth bringing this up for discussion in the community. Here are the details of the issue:

### Background

We used `spring-kafka` with version 2.3.13.RELEASE, which depends on `kafka-clients` version 2.5.1.

In the application, we use `DefaultKafkaProducerFactory` to create the Producer and set it to per-thread mode, meaning we allocate a producer instance for each thread in the worker thread group. Here is core codes:

```
@Bean
@Primary
public KafkaTemplate<String, String> kafkaTemplate() {
    return new KafkaTemplate<>(producerFactory());
}

@Bean
public ProducerFactory<String, String> producerFactory() {
    DefaultKafkaProducerFactory<String, String> factory = new DefaultKafkaProducerFactory<>(producerConfigs());
    factory.setProducerPerThread(true);
    return factory;
}

public Map<String, Object> producerConfigs() {
    Map<String, Object> props = new HashMap<>();
    props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, servers);
    props.put(ProducerConfig.ACKS_CONFIG, acks);
    props.put(ProducerConfig.RETRIES_CONFIG, retries);
    props.put(ProducerConfig.BATCH_SIZE_CONFIG, batchSize);
    props.put(ProducerConfig.LINGER_MS_CONFIG, linger);
    props.put(ProducerConfig.BUFFER_MEMORY_CONFIG, bufferMemory);
    props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
    props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
    props.put(ProducerConfig.MAX_REQUEST_SIZE_CONFIG, maxRequestSize);
    props.put(ProducerConfig.COMPRESSION_TYPE_CONFIG, compressionType);
    props.put(ProducerConfig.SEND_BUFFER_CONFIG, sendBuffer);
    KafkaSaslConfig.kafkaSaslConfig(props, propertiesConfig);
    return props;
}
```

When we install the Java Agent in production and start the application, it throws an exception: 
```
org.apache.kafka.common.config.ConfigException: Missing required configuration "key.serializer" which has no default value.
```
However, the `key.serializer` is indeed present in our actual code as you can see above.

### Problem Localization

The application works fine when run independently. But when it runs with OTel Java Agent, the actual configuration only includes:
```
@HashMap[
    @String[compression.type]:@String[gzip],
    @String[metric.reporters]:@String[io.opentelemetry.javaagent.shaded.instrumentation.kafka.internal.OpenTelemetryMetricsReporter],
    @String[acks]:@String[-1],
    @String[batch.size]:@Integer[4096],
    @String[key.serializer]:@Class[class org.apache.kafka.common.serialization.StringSerializer],     @String[opentelemetry.supplier]:@OpenTelemetrySupplier[io.opentelemetry.javaagent.shaded.instrumentation.kafka.internal.OpenTelemetrySupplier@20db01b6],
    @String[opentelemetry.instrumentation_name]:@String[io.opentelemetry.kafka-clients-0.11],
    @String[retries]:@Integer[1],
    @String[max.request.size]:@Integer[2097152],
    @String[value.serializer]:@Class[class org.apache.kafka.common.serialization.StringSerializer],
]
```
Since our configuration has exactly 11 items, very close to the threshold for automatic resizing of HashMap, we suspect that concurrently adding keys to a resizing HashMap caused some nodes to be lost.

We reviewed the code of the OTel Java Agent and `spring-kafka` and we confirmed that this is the cause:

In `spring-kafka`, when creating `DefaultKafkaProducerFactory`, the config is stored in a HashMap. Later, when creating the producer, it directly reuses this config, even in a multi-threaded environment (which is fine if the config does not change).
<img width="804" alt="截屏2024-10-30 22 12 19" src="https://github.com/user-attachments/assets/ace2d254-4626-4135-b8f5-7d84a3cf5910">
<img width="641" alt="截屏2024-10-30 22 13 07" src="https://github.com/user-attachments/assets/49746d12-e496-48d1-b8bf-dce0570301ef">

When creating the producer, it first enters the OTel Java Agent's advice code block, which executes the `enhanceConfig()` method. During this process, the config map is "unexpectedly" modified concurrently, leading to concurrency conflicts.
https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/00de5e3016308428b08c9bf4546d10ae2986125b/instrumentation/kafka/kafka-clients/kafka-clients-0.11/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/kafkaclients/v0_11/metrics/KafkaMetricsProducerInstrumentation.java#L42-L50

### Fix

By creating a copy of the incoming config before calling `enhanceConfig()`, the issue was resolved.
```
 @Advice.OnMethodEnter(suppress = Throwable.class) 
 public static void onEnter( 
         @Advice.Argument(value = 0, readOnly = false) Map<String, Object> config) { 
     config = new HashMap<>(config);
     enhanceConfig(config); 
 } 
```

### Some Thoughts

Similar to another issue #11946  I have raised, it seems that Java Agent developers are often troubled by such corner cases. Is there any approaches or suggestions to anticipate these kinds of issues rather than solving them after they occur?

**Comment by trask** [comment]
> By creating a copy of the incoming config before calling `enhanceConfig()`, the issue was resolved.

can you send a PR for this?

> Is there any approaches or suggestions to anticipate these kinds of issues rather than solving them after they occur?

there are some kinds of issues that we only find out through tests or user reports. luckily we seem to have a lot of users, and so we hear about these kinds of issues once and can fix them for everyone

**Comment by Cirilla-zmh** [comment]
> can you send a PR for this?

Glad to do so! It will take me some time to construct a test case.

**Comment by trask** [comment]
> It will take me some time to construct a test case.

if it only reproduces under a race condition, I wouldn't worry about a test case, and just document the code well

**Comment by Cirilla-zmh** [comment]
@trask 
Hi, I have already submitted the PR.
To add, the issue was fixed in `spring-kafka` version 2.5.0.RC1—although as a solution to other problems. Here is the [commit](https://github.com/spring-projects/spring-kafka/commit/8347b957b75ce7c81d1bdaedc807f62f543dbeae#diff-ee2dfb7123c02f066d237b437f06f37c6ab26279ed47f3edd8b7ca0d0a4c96cd).
